geode-apis-compatible-with-redis/src/distributedTest/java/org/apache/geode/redis/internal/executor/string/MSetNXDUnitTest.java

```diff
@@ -0,0 +1,174 @@
+import org.apache.geode.test.dunit.rules.MemberVM;
+import org.apache.geode.test.dunit.rules.RedisClusterStartupRule;
+import org.apache.geode.test.junit.rules.ExecutorServiceRule;
+
+public class MSetNXDUnitTest {
+
+  @ClassRule
+  public static RedisClusterStartupRule clusterStartUp = new RedisClusterStartupRule(4);
+
+  @ClassRule
+  public static ExecutorServiceRule executor = new ExecutorServiceRule();
+
+  private static final String HASHTAG = "{tag}";
+  private static JedisCluster jedis;
+  private static MemberVM server1;
+  private static int locatorPort;
+
+  private static MemberVM locator;
+
+  @BeforeClass
+  public static void classSetup() {
+    locator = clusterStartUp.startLocatorVM(0);
+    locatorPort = locator.getPort();
+    server1 = clusterStartUp.startRedisVM(1, locatorPort);
+    clusterStartUp.startRedisVM(2, locatorPort);
+    clusterStartUp.startRedisVM(3, locatorPort);
+
+    int redisServerPort1 = clusterStartUp.getRedisPort(1);
+    jedis = new JedisCluster(new HostAndPort(BIND_ADDRESS, redisServerPort1), REDIS_CLIENT_TIMEOUT);
+  }
+
+  @After
+  public void after() {
+    clusterStartUp.flushAll();
+  }
+
+  @AfterClass
+  public static void tearDown() {
+    jedis.close();
+  }
+
+  @Test
+  public void testMSetnx_concurrentInstancesHandleBucketMovement() {
+    int KEY_COUNT = 5000;
+    String[] keys = new String[KEY_COUNT];
+
+    for (int i = 0; i < keys.length; i++) {
+      keys[i] = HASHTAG + "key" + i;
+    }
+    String[] keysAndValues1 = makeKeysAndValues(keys, "valueOne");
+    String[] keysAndValues2 = makeKeysAndValues(keys, "valueTwo");
+
+    new ConcurrentLoopingThreads(100,
+        i -> jedis.msetnx(keysAndValues1),
+        i -> jedis.msetnx(keysAndValues2),
+        i -> clusterStartUp.moveBucketForKey(keys[0]))
+            .runWithAction(() -> {
+              assertThat(jedis.mget(keys)).satisfiesAnyOf(
+                  values -> assertThat(values)
+                      .allSatisfy(value -> assertThat(value).startsWith("valueOne")),
+                  values -> assertThat(values)
+                      .allSatisfy(value -> assertThat(value).startsWith("valueTwo")));
+              jedis.del(keys);
+            });
+  }
+
+  @Ignore("GEODE-9604")
+  @Test
+  public void testMSetnx_crashDoesNotLeaveInconsistencies() throws Exception {
+    int KEY_COUNT = 1000;
+    String[] keys = new String[KEY_COUNT];
+
+    for (int i = 0; i < keys.length; i++) {
+      keys[i] = HASHTAG + "key" + i;
+    }
+    String[] keysAndValues1 = makeKeysAndValues(keys, "valueOne");
+    String[] keysAndValues2 = makeKeysAndValues(keys, "valueTwo");
+    AtomicBoolean running = new AtomicBoolean(true);
+
+    Future<?> future = executor.submit(() -> {
+      for (int i = 0; i < 20 && running.get(); i++) {
+        clusterStartUp.moveBucketForKey(keys[0], "server-3");
+        // Sleep for a bit so that MSETs can execute
+        Thread.sleep(2000);
```

Comment:
Another option that can be a bit cleaner and more efficient is to introduce a `CountDownLatch` above this block. All of these threads would then await the latch:
```
CountDownLatch latch = new CountDownLatch(1);
```
```
Future<?> future = executor.submit(() -> {
      for (int i = 0; i < 20 && running.get(); i++) {
        clusterStartUp.moveBucketForKey(keys[0], "server-3");
        // Sleep for a bit so that MSETs can execute
        latch.await(GeodeAwaitility.getTimeout().getSeconds(), TimeUnit.SECONDS);
```
And then later on where you're executing MSETs or whatever you want to sleep a bit for, you open that latch:
```
latch.countDown();
```
If there are _n_ threads concurrently doing something you want to pause for, then use `new CountDownLatch(n)` and have each of those threads `countDown` when you reach the mid-point or whatever you're pausing for.

Reply:
Yeah, I was looking sideways at that sleep also, but in the interest in getting this ticket out of the critical path, I'm inclined to leave this for now. If you think this is important enough that it should be changed, can you open a ticket for it? Thanks.